lib/foreman_maintain/runner.rb

```diff
@@ -1,48 +1,73 @@
 module ForemanMaintain
   # Class responsible for running the scenario
   class Runner
+    attr_reader :reporter, :exit_code
+
     require 'foreman_maintain/runner/execution'
+    require 'foreman_maintain/runner/stored_execution'
     def initialize(reporter, scenarios, options = {})
-      options.validate_options!(:assumeyes)
+      options.validate_options!(:assumeyes, :whitelist, :force)
       @assumeyes = options.fetch(:assumeyes, false)
+      @whitelist = options.fetch(:whitelist, [])
+      @force = options.fetch(:force, false)
       @reporter = reporter
       @scenarios = Array(scenarios)
-      @scenarios_with_dependencies = scenarios_with_dependencies
       @quit = false
+      @last_scenario = nil
+      @exit_code = 0
     end
 
-    def assumeyes?
-      @assumeyes
+    def quit?
+      @quit
     end
 
-    def scenarios_with_dependencies
-      @scenarios.map do |scenario|
-        scenario.before_scenarios + [scenario]
-      end.flatten
+    def assumeyes?
+      @assumeyes
     end
 
     def run
-      scenarios_with_dependencies.each do |scenario|
+      @scenarios.each do |scenario|
         run_scenario(scenario)
+        break if @quit
       end
     end
 
-    def run_scenario(scenario)
-      @steps_to_run = scenario.steps.dup
-      @steps_to_run = ForemanMaintain::DependencyGraph.sort(@steps_to_run)
-      @reporter.before_scenario_starts(scenario)
-      while !@quit && !@steps_to_run.empty?
-        step = @steps_to_run.shift
-        @reporter.puts('Rerunning the check after fix procedure') if rerun_check?(step)
-        execution = Execution.new(step, @reporter)
-        execution.run
-        ask_about_offered_steps(step)
+    # rubocop:disable Metrics/MethodLength, Metrics/AbcSize, Metrics/CyclomaticComplexity
+    def run_scenario(scenario, confirm = true)
+      return if scenario.steps.empty?
+      raise 'The runner is already in quit state' if quit?
```

Comment:
This is more of a internal developer assertion, that we should not get into, unless there is some bug